// @haxtheweb/rpg-character/test/rpg-contributors.js
import { LitElement, html, css } from 'lit';
import { RpgCharacter } from './rpg-character.js'; 

export class RpgContributors extends LitElement {
  static get tag() {
    return 'github-rpg-contributors';
  }

  constructor() {
    super();
    this.organization = 'haxtheweb'; 
    this.repo = 'webcomponents'; 
    this.limit = 6; 
    this.contributors = []; 
  }

  static get properties() {
    return {
      organization: { type: String },
      repo: { type: String },
      limit: { type: Number },
      contributors: { type: Array },
    };
  }

  connectedCallback() {
    super.connectedCallback();
    this.fetchContributors(); 
  }

  async fetchContributors() {
    try {
      const url = `https://api.github.com/repos/${this.organization}/${this.repo}/contributors`;
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error('Failed to fetch contributors');
      }
      const data = await response.json();
      this.contributors = data.slice(0, this.limit); 
      this.requestUpdate(); 
    } catch (error) {
      console.error('Error fetching contributors:', error);
    }
  }

  static get styles() {
    return css`
      :host {
        display: block;
        padding: 16px;
      }

      .contributors {
        display: flex;
        flex-wrap: wrap;
        gap: 20px;
      }

      .contributor {
        text-align: center;
        border: 1px solid #ddd;
        padding: 10px;
        border-radius: 8px;
        background-color: #f9f9f9;
        box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
      }

      rpg-character {
        cursor: pointer;
        display: block;
        margin: 0 auto;
      }

      a {
        text-decoration: none;
        color: inherit;
      }

      p {
        margin: 5px 0;
        font-size: 14px;
      }
    `;
  }

  render() {
    return html`
      <div class="contributors">
        ${this.contributors.map(
          (contributor) => html`
            <div class="contributor">
              <a href="${contributor.html_url}" target="_blank" rel="noopener noreferrer">
                <rpg-character seed="${contributor.login}"></rpg-character>
              </a>
              <p>${contributor.login}</p>
              <p>Contributions: ${contributor.contributions}</p>
            </div>
          `
        )}
      </div>
    `;
  }
}

globalThis.customElements.define(GitHubRpgContributors.tag, GitHubRpgContributors);
